server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java

```diff
@@ -573,12 +578,43 @@ public void addAllocFailuresResetListenerTo(ClusterService clusterService) {
         });
 
         clusterService.addListener((changeEvent) -> {
-            if (changeEvent.nodesAdded() && changeEvent.state().getRoutingNodes().hasAllocationFailures()) {
+            if (shouldResetAllocationFailures(changeEvent)) {
                 taskQueue.submitTask("reset-allocation-failures", (e) -> { assert MasterService.isPublishFailureException(e); }, null);
             }
         });
     }
 
+    /**
+     *  We should reset allocation/relocation failure count to allow further retries when:
```

Comment:
In the PR, we essentially reset when we know there might be shard movements caused by the events chosen here. For that reason, I'm not considering RESTART. Same logic as in https://github.com/elastic/elasticsearch/pull/106998.

Reply:
I had an unsuccessful attempt at "generalizing" that logic to reuse here, but it ended up a bit weird. So I've just reused those few lines. When we reset the desired balance, and when we reset the allocation/relocation attempts, might not necessarily always share the same logic, I suppose, so not bad to have them separate IMO.